contracts/HookedTokenManager.sol

```diff
@@ -6,23 +6,20 @@
 
 pragma solidity 0.4.24;
 
-import "@aragon/os/contracts/apps/AragonApp.sol";
-import "@aragon/os/contracts/common/IForwarder.sol";
+import "@aragon/apps-token-manager/contracts/TokenManager.sol";
 
-import "@aragon/os/contracts/lib/math/SafeMath.sol";
+import "./TokenManagerHook.sol";
 
-import "@aragon/apps-shared-minime/contracts/ITokenController.sol";
-import "@aragon/apps-shared-minime/contracts/MiniMeToken.sol";
 
-
-contract TokenManager is ITokenController, IForwarder, AragonApp {
+contract HookedTokenManager is TokenManager {
```

Comment:
I was not sure how polymorphism in Solidity works. I would like to use `HookedTokenManager` as a `TokenManager` most of the times, and I don't know if it is possible if the child contract is not present in the inheritance tree.

I tried by only overriding the functions that are being modified, but then the Aragon Client was not responding properly, maybe because the artifact was not detecting well the inherited external calls (ganache tests were working perfectly).

Do you know if it's safe to cast a `HookedTokenManager` into a `TokenManager` even if one does not inherit from the other? And on the other hand, do you know if overriding functions has a negative impact on gas needed? What is more reasonable? Should we fix the `artifact.json` so it gets all functions properly?

And there is that other problem with mintings from the new TM interface I'm still trying to fix.

Reply:
> I was not sure how polymorphism in Solidity works. I would like to use `HookedTokenManager` as a `TokenManager` most of the times, and I don't know if it is possible if the child contract is not present in the inheritance tree.

You can cast any address to a `TokenManager` to pass to whatever function.

> I tried by only overriding the functions that are being modified, but then the Aragon Client was not responding properly, maybe because the artifact was not detecting well the inherited external calls (ganache tests were working perfectly).

Inherited contracts interfaces weren't available in the past to the client API, it sounds like they still aren't. I don't know whether it should inherit TokenManager or not but this is another reason why it shouldn't, as it will be empty.

> Do you know if it's safe to cast a `HookedTokenManager` into a `TokenManager` even if one does not inherit from the other?

It is as long as the interface of the TokenManager is maintained.

> And on the other hand, do you know if overriding functions has a negative impact on gas needed?

I doubt it, afaik the Solidity compiler copies all the functions referenced from the parent contract into one contract and compiles that.

>  Should we fix the artifact.json so it gets all functions properly?

This is the ideal solution. Although I think it might work once we migrate to Buidler and use Truffle v5: https://github.com/aragon/aragon-cli/issues/133

Reply:
Very good. I think we are ready to go as we are, when Token Manager is migrated to truffle v5 we can consider inherit from the contract and reduce the number of overriden functions.

Reply:
In hindsight I think it's reasonable for it to inherit the TokenManager, since it would be useful to not have to cast it as such when passing it to functions that require a TokenManager. Just a shame we have to overwrite all the functionality in it.